src/Microsoft.OpenApi.Readers/Services/OpenApiReferenceResolver.cs

```diff
@@ -0,0 +1,207 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. 
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.OpenApi.Interfaces;
+using Microsoft.OpenApi.Models;
+using Microsoft.OpenApi.Services;
+
+namespace Microsoft.OpenApi.Readers.Services
+{
+    /// <summary>
+    /// This class is used to walk an OpenApiDocument and convert unresolved references to references to populated objects
+    /// </summary>
+    internal class OpenApiReferenceResolver : OpenApiVisitorBase
+    {
+        private OpenApiDocument _currentDocument;
+        private bool _resolveRemoteReferences;
+
+        public OpenApiReferenceResolver(OpenApiDocument currentDocument, bool resolveRemoteReferences = true) 
+        {
+            _currentDocument = currentDocument;
+            _resolveRemoteReferences = resolveRemoteReferences;
+        }
+
+        public override void Visit(OpenApiDocument doc)
+        {
+            if (doc.Tags != null)
+            {
+                ResolveTags(doc.Tags);
+            }
+        }
+
+        public override void Visit(OpenApiComponents components)
+        {
+            ResolveMap(components.Parameters);
+            ResolveMap(components.RequestBodies);
+            ResolveMap(components.Responses);
+            ResolveMap(components.Links);
+            ResolveMap(components.Callbacks);
+            ResolveMap(components.Examples);
+            ResolveMap(components.Schemas);
+            ResolveMap(components.SecuritySchemes);
```

Comment:
Shouldn't it be the job of the walker to do this "walking"? 

Shouldn't this file just do reference-resolving for the object that's actually a reference and then leave the logic of getting into that object to the Walker?

Reply:
I wish I could do that.  However, we replace the "unresolved" object that is empty, except for the Reference, with the completely filled object.  Therefore, I need to detect the reference from the parent object that holds the pointer to the unresolved object so that we can change it.

Reply:
That is true. It's kinda unfortunate since that means we can't fully leverage the Walking capability. 

If we revise the Walk method design to take in another variable (in this case the parent object), and then it can make changes to that variable, then that would work. But that's a big change, and it doesn't seem clean.

Reply:
@PerthCharern Yes, the walker is not ideal in this scenario.  However, it is expected to use the parent object to do certain types of work.  e.g. Checking if a value is not null, needs to be done from above.  Checking if there are no duplicate parameter names.  Checking if two peer values are not contradicting each other.

Passing in the parent object is hard because you don't know its type.  It would need to be passed in as an object and then cast to each of the possible parent types.  In the end I think it is just as easy to do certain contextually sensitive work from the parent visit method.